File: waterproof/src/main/java/mx/yellowme/waterproof/mvp/lists/ListAdapter.java
package mx.yellowme.waterproof.mvp.lists;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;

import java.util.List;

import mx.yellowme.waterproof.mvp.BaseAdapter;

import static mx.yellowme.waterproof.WaterproofConditions.checkNotNull;

/**
 * Base class to implement a RecyclerView based list. In order to use this class
 * is necessary to implement a Custom ViewHolder over which the adapter is going to operate.
 */
public abstract class ListAdapter<Item, CustomViewHolder extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<CustomViewHolder>
        implements BaseAdapter {

    protected List<Item> mItemList;
    protected Context mContext;
    protected ItemListener<Item> mItemListener;

    protected ListAdapter(@NonNull List<Item> itemList, @NonNull ItemListener<Item> itemListener){
        setList(itemList);
        mItemListener = checkNotNull(itemListener);
    }

    /**
     * Replaces all the objects of the current list on adapter.
     * @param users list of object to replaced the current list.
     */
    public void replaceData(List<Item> users) {
        setList(users);
        notifyDataSetChanged();
    }

    /**
     * Adds more objects to current list on adapter.
     * @param users list of new objects to be added to the current list.
     */
    public void addData(List<Item> users) {
        addToList(users);
        notifyDataSetChanged();
    }


    @Override
    public int getItemCount() {
        return mItemList.size();
    }

    public Item getItem(int position) {
        return mItemList.get(position);
    }

    /**
     * Helper method to interact with the item list variable.
     */
    protected void setList(List<Item> elements) {
        mItemList = checkNotNull(elements);
    }

    /**
     * Helper method to interact with the item list variable.
     */
    protected void addToList(List<Item> elements) {
        checkNotNull(elements);
        mItemList.addAll(elements);
    }

    public List<Item> getItems() {
        return mItemList;
    }
}

File: waterproof/src/main/java/mx/yellowme/waterproof/common/adapter/ListAdapter.kt
package mx.yellowme.waterproof.common.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.animation.DecelerateInterpolator
import mx.yellowme.waterproof.common.utils.Utils

/**
 * Base class to implement a RecyclerView based list. In order to use this class
 * is necessary to implement a Custom ViewHolder over which the adapter is going to operate.
 */
abstract class ListAdapter<Item, ConcreteViewHolder : BindableViewHolder<Item>>(
        val context: Context,
        val listener: (Item) -> Unit
) : RecyclerView.Adapter<ConcreteViewHolder>() {

    /**
     * Helper for running animations when loading
     */
    private var lastAnimatedPosition = -1

    var items: List<Item>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    open var animateEnter: Boolean = false

    override fun onBindViewHolder(holder: ConcreteViewHolder, position: Int) {
        if(animateEnter) { runEnterAnimation(holder.itemView, position) }

        items?.let {
            val current = it[position]
            (holder as? BindableViewHolder<Item>)?.apply {
                bind(current)
            }
        }
    }

    /**
     * Replaces all the objects of the current list on adapter.
     * @param elements list of object to replaced the current list.
     */
    fun replaceData(elements: List<Item>) {
        setList(elements)
    }

    /**
     * Adds more objects to current list on adapter.
     * @param elements list of new objects to be added to the current list.
     */
    fun addData(elements: List<Item>) {
        addToList(elements)
    }

    /**
     * Total elements on the adapter
     */
    override fun getItemCount(): Int {
        return items?.size?: 0
    }

    /**
     * Helper method to interact with the item list variable.
     */
    open fun setList(elements: List<Item>) {
        addToList(elements)
    }

    /**
     * Helper method to interact with the item list variable.
     */
    open fun addToList(elements: List<Item>) {
        items = elements
    }

    private fun runEnterAnimation(view: View, position: Int) {
        if (position > lastAnimatedPosition) {
            lastAnimatedPosition = position
            view.translationY = Utils.getScreenHeight(context);
            view.animate()
                    .translationY(0f)
                    .setInterpolator(DecelerateInterpolator(3f))
                    .setDuration(700)
                    .start()
        }
    }
}
